File: src/main/java/de/zalando/zally/rules/UseProblemJsonRule.java
package de.zalando.zally.rules;

import de.zalando.zally.Violation;
import de.zalando.zally.ViolationType;
import io.swagger.models.HttpMethod;
import io.swagger.models.Operation;
import io.swagger.models.Path;
import io.swagger.models.Response;
import io.swagger.models.Swagger;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;


public class UseProblemJsonRule implements Rule {
    private static final String TITLE = "Use Problem JSON";
    private static final String DESCRIPTION = "Operations Should Return Problem JSON When Any Problem Occurs During Processing Whether Caused by Client Or Server";
    private static final ViolationType VIOLATION_TYPE = ViolationType.MUST;
    private static final String RULE_LINK = "https://zalando.github.io/restful-api-guidelines/compatibility/Compatibility.html#must-use-problem-json";

    @Override
    public List<Violation> validate(Swagger swagger) {
        List<Violation> violations = new ArrayList<>();
        if (swagger == null || swagger.getPaths() == null) {
            return violations;
        }
        for(Map.Entry<String, Path> path: swagger.getPaths().entrySet())
            for(Map.Entry<HttpMethod, Operation> op : path.getValue().getOperationMap().entrySet())
                for(Map.Entry<String, Response> response: op.getValue().getResponses().entrySet()) {
                    Integer httpCodeInt = null;
                    try {
                        httpCodeInt = Integer.valueOf(response.getKey());
                    } catch (NumberFormatException nfe) {
                        // ignore any http codes which are not between 400 and 599
                    }
                    //TODO: check if schema reference type is Problem := https://github.com/swagger-api/swagger-parser/pull/303/files
                    if (httpCodeInt != null && httpCodeInt >= 400 && httpCodeInt <= 599
                            && response.getValue().getSchema() != null
                            && !"ref".equals(response.getValue().getSchema().getType())) {
                        String violationPath = path.getKey() + "/" + op.getKey() + "/" + response.getKey();
                        violations.add(new Violation(TITLE, DESCRIPTION, VIOLATION_TYPE, RULE_LINK, violationPath));
                    }
                }
        return violations;
    }

}

File: server/src/main/java/de/zalando/zally/rule/UseProblemJsonRule.kt
package de.zalando.zally.rule

import de.zalando.zally.dto.Violation
import de.zalando.zally.dto.ViolationType
import io.swagger.models.ComposedModel
import io.swagger.models.HttpMethod
import io.swagger.models.Model
import io.swagger.models.Operation
import io.swagger.models.RefModel
import io.swagger.models.Response
import io.swagger.models.Swagger
import io.swagger.models.properties.ObjectProperty
import io.swagger.models.properties.RefProperty
import org.springframework.stereotype.Component

@Component
class UseProblemJsonRule : AbstractRule() {
    override val title = "Use Problem JSON"
    override val url = "https://zalando.github.io/restful-api-guidelines/common-data-objects/CommonDataObjects.html" +
        "#must-use-problem-json"
    override val violationType = ViolationType.MUST
    override val code = "M015"
    private val description = "Operations Should Return Problem JSON When Any Problem Occurs During Processing " +
        "Whether Caused by Client Or Server"
    private val requiredFields = setOf("title", "status")

    override fun validate(swagger: Swagger): Violation? {
        val paths = swagger.paths.orEmpty().flatMap { pathEntry ->
            pathEntry.value.operationMap.orEmpty().filter { it.key.shouldContainPayload() }.flatMap { opEntry ->
                opEntry.value.responses.orEmpty().flatMap { responseEntry ->
                    val httpCode = responseEntry.key.toIntOrNull()
                    if (httpCode in 400..599 && !isValidProblemJson(swagger, responseEntry.value, opEntry.value)) {
                        listOf("${pathEntry.key} ${opEntry.key} ${responseEntry.key}")
                    } else emptyList()
                }
            }
        }

        return if (paths.isNotEmpty()) Violation(this, title, description, violationType, url, paths) else null
    }

    private fun isValidProblemJson(swagger: Swagger, response: Response, operation: Operation) =
        isProblemJson(swagger, response) && producesJson(swagger, operation)

    private fun isProblemJson(swagger: Swagger, response: Response): Boolean {
        val schema = response.schema
        val properties = when (schema) {
            is RefProperty -> getProperties(swagger, swagger.definitions[(response.schema as RefProperty).simpleRef])
            is ObjectProperty -> schema.properties?.keys.orEmpty()
            else -> emptySet<String>()
        }
        return properties.containsAll(requiredFields)
    }

    private fun getProperties(swagger: Swagger, definition: Model?): Set<String> {
        return when (definition) {
            is ComposedModel -> definition.allOf.orEmpty().flatMap { getProperties(swagger, it) }.toSet()
            is RefModel -> getProperties(swagger, swagger.definitions[definition.simpleRef])
            else -> definition?.properties?.keys.orEmpty()
        }
    }

    private fun producesJson(swagger: Swagger, operation: Operation) =
        if (operation.produces.orEmpty().isEmpty()) {
            swagger.produces.orEmpty().containsJson()
        } else {
            operation.produces.containsJson()
        }

    // support for application/json also with set charset e.g. "application/json; charset=utf-8"
    private fun List<String>.containsJson() =
        any { it.startsWith("application/json") }

    private fun HttpMethod.shouldContainPayload(): Boolean =
        name.toLowerCase() !in listOf("head", "options")

}
